qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,213 @@
+            regname = 'meas'
+            circ = self._pe_circuit
+            if not regname in [reg.name for reg in circ.cregs]:
+                creg = ClassicalRegister(self._num_evaluation_qubits, regname)
+                circ.add_register(creg)
+                circ.barrier()
+                circ.measure(range(self._num_evaluation_qubits), range(self._num_evaluation_qubits))
+
+    @property
+    def phases(self):
+        """
+        Returns the all phases and their frequencies. This is either a dict whose
+        keys are bit strings and values are counts, or an array whose values correspond
+        to weights on bit strings.
+        """
+        return self._phases
+
+    @property
+    def pe_circuit(self):
+        """ Return the phase estimation circuit. """
+        return self._pe_circuit
+
+    def _compute_phases(self):
+        if self._quantum_instance.is_statevector:
+            state_vec = self._result.get_statevector()
+            evaluation_density_matrix = get_subsystem_density_matrix(
+                state_vec,
+                range(self._num_evaluation_qubits, self._num_evaluation_qubits + self._num_unitary_qubits)
+            )
+            phases = evaluation_density_matrix.diagonal().real # The diagonal is real
+        else:
+            # return counts with keys sorted numerically
+            counts = self._result.get_counts()
+            phases = {k[::-1] : counts[k] for k in counts.keys()}
+            phases = _sort_phases(phases)
+            phases = qiskit.result.Counts(phases, memory_slots=counts.memory_slots, creg_sizes=counts.creg_sizes)
+
+        self._phases = phases
+
+    # If we reversed the bit order of evaluation register circuit (including the iqft) then we would avoid
+    # the binary calculations in the case of the statevector simulator.
+    def single_phase(self):
+        """
+        Return the estimated phase as a number between 0.0 and 1.0, with 1.0 corresponding to a
+        phase of 2pi. It is assumed that the input vector is an eigenvecter of the unitary so that
+        the peak of the probability density occurs at the bit string that most closely approximates
+        the true phase.
+        """
+        if self._quantum_instance.is_statevector:
+            idx = np.argmax(abs(self._phases)) # np.argmax ignores complex part of number. But, we take abs anyway
+            binary_phase_string = np.binary_repr(idx, self._num_evaluation_qubits)[::-1]
+
+        else:
+            binary_phase_string = max(self._phases, key=self._phases.get)
+
+        phase = bit_string_to_phase(binary_phase_string)
+        return phase
+
+    def _run(self):
+        """
+        Run the circuit and return the estimated phase as a number between 0.0 and 1.0, with 1.0 corresponding
+        to a phase of 2pi.
+        """
+
+        result = self._quantum_instance.execute(self._pe_circuit)
+        self._result = result
+        self._compute_phases()
```

Comment:
This method should return a Results object, can you add a Results class for phase estimation algorithms?

Reply:
Probably a good idea.